jena-fuseki2/jena-fuseki-webapp/src/main/java/org/apache/jena/fuseki/mgt/ActionCompact.java

```diff
@@ -0,0 +1,69 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.jena.fuseki.mgt;
+
+import static java.lang.String.format;
+
+import org.apache.jena.fuseki.ctl.ActionAsyncTask;
+import org.apache.jena.fuseki.ctl.TaskBase;
+import org.apache.jena.fuseki.servlets.HttpAction;
+import org.apache.jena.fuseki.servlets.ServletOps;
+import org.apache.jena.tdb2.DatabaseMgr;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ActionCompact extends ActionAsyncTask
+{
+    public ActionCompact() { super("Compact"); }
+
+    @Override
+    public void validate(HttpAction action) {}
+
+    @Override
+    protected Runnable createRunnable(HttpAction action) {
+        String name = getItemName(action);
+        if ( name == null ) {
+            action.log.error("Null for dataset name in item request");
+            ServletOps.errorOccurred("Null for dataset name in item request");
+            return null;
+        }
+
```

Comment:
Maybe test it is the right species of DatrasetGraph and return an error if it is not.

See `TDBOps` in fuseki-core.

Reply:
`DatabaseMgr.compact()` already does this checking so didn't seem relevant to duplicate that here.  Either the dataset can be compacted or calling `compact()` will throw a relevant error

Reply:
It throws `TDBException("Not a switchable TDB database");` so won't that mean it is "500 internal error" not "400 Bad request".

Reply:
Well that call is going to happen on the background task thread and any failure will just get logged and not bubble up AFAICT

Reply:
If the test is in "validate", it is in the request cycle.

The contract of `ActionAsyncTask` and `run()` could do with some clearing up (separate to this PR!). Looks a bit messy / quick hackish. e.g. The `TaskBase` constructor have args for the element of the `HttpAction` rather than the implementation comment + single `HttpAction` arg.

Reply:
Yeah there's definitely room for some clean up I was trying to figure out a way to add a success/failure flag to tasks but with the multiple layers of wrapping couldn't find a reliable way to get this working

`AsyncPool.submit()` creates a `Callable<Object>`, that wraps a `Runnable`, and then creates and submits an `AsyncTask` which is itself the `Callable<Object>` that actually gets submitted to the executor service

Reply:
Figured this out with the latest commit, various parts of the async task code was just logging an error and continuing on rather than throwing it upwards which prevented detecting the error at the point where we could set a success flag appropriately.  Unit tests are updated to check for this flag as appropriate

Reply:
If this is done now, could you resolve the ticket? Thanks.